messaging/src/main/java/org/axonframework/eventhandling/deadletter/EvaluationTask.java

```diff
@@ -0,0 +1,130 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.eventhandling.deadletter;
+
+import org.axonframework.common.transaction.TransactionManager;
+import org.axonframework.eventhandling.EventMessage;
+import org.axonframework.eventhandling.EventMessageHandler;
+import org.axonframework.eventhandling.ListenerInvocationErrorHandler;
+import org.axonframework.messaging.deadletter.DeadLetter;
+import org.axonframework.messaging.deadletter.DeadLetterEvaluationException;
+import org.axonframework.messaging.deadletter.DeadLetterQueue;
+import org.axonframework.messaging.unitofwork.DefaultUnitOfWork;
+import org.axonframework.messaging.unitofwork.UnitOfWork;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A {@link Runnable} implementation used to evaluate a {@link DeadLetter} taken from the {@link DeadLetterQueue}. This
+ * task is added through {@link DeadLetterQueue#onAvailable(String, Runnable)}, so we can typically assume there are
+ * entries ready for evaluation.
+ *
+ * @author Steven van Beelen
+ * @since 4.6.0
+ */
+class EvaluationTask implements Runnable {
+
+    private final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final List<EventMessageHandler> eventHandlingComponents;
+    private final DeadLetterQueue<EventMessage<?>> queue;
+    private final String processingGroup;
+    private final TransactionManager transactionManager;
+    private final ListenerInvocationErrorHandler listenerInvocationErrorHandler;
+
+    EvaluationTask(List<EventMessageHandler> eventHandlingComponents,
+                   DeadLetterQueue<EventMessage<?>> queue,
+                   String processingGroup,
+                   TransactionManager transactionManager,
+                   ListenerInvocationErrorHandler listenerInvocationErrorHandler) {
+        this.eventHandlingComponents = eventHandlingComponents;
+        this.queue = queue;
+        this.processingGroup = processingGroup;
+        this.transactionManager = transactionManager;
+        this.listenerInvocationErrorHandler = listenerInvocationErrorHandler;
+    }
+
+    @Override
+    public void run() {
+        AtomicBoolean evaluationFailed = new AtomicBoolean(false);
+        Optional<DeadLetter<EventMessage<?>>> optionalLetter;
+        while (!evaluationFailed.get() &&
+                (optionalLetter = transactionManager.fetchInTransaction(() -> queue.take(processingGroup))).isPresent()
```

Comment:
I do, take a look at my JPA branch. 
![image](https://user-images.githubusercontent.com/3278878/174743668-bbaad039-8d09-4513-9c6e-bf66166c4dfa.png)


Also, don't forget to add the catch-block like I have here. Exceptions can fall out of the thread otherwise, as I have experienced before

Reply:
Awesome recommendation @Morlack. I'm gonna copy that and give you author creds in this file.